### PR TITLE
[BUILD-881] Reconciles SharedResource

### PR DESCRIPTION
**Pull request by apoorvajagtap**
This commit adds the logic to BootStrap SharedResource on the creation of OpenShiftBuild CR, and the add-on steps depending on the Enabled/Disabled state of SharedResource.

### CODE REVIEW

File: internal/controller/sharedresource_controller.go
Comment (sayan-biswas) [inline internal/controller/sharedresource_controller.go:89]: There's already a controller for this resource. You can call the shared resource login in the Reconcile method of that controller. Since we are not creating new CRD this is not required.

File: config/sharedresource/shared_resource_privileged_scc.yaml
Comment (adambkaplan) [inline config/sharedresource/shared_resource_privileged_scc.yaml:1]: Not absolutely necessary for GA - we can use the `privileged` SCC that ships with OpenShift by default.

All other GA CSI drivers use the `privileged` SCC. See https://github.com/openshift/aws-efs-csi-driver-operator/blob/master/assets/rbac/privileged_role.yaml

File: internal/sharedresource/sharedresource.go
Comment (adambkaplan) [inline internal/sharedresource/sharedresource.go:1]: We should have some level of testing defined for the operator logic - either via EnvTest with a suite for `openshiftbuild_controller`, or here and test with controller-runtime fake clients.

File: internal/sharedresource/sharedresource.go
Comment (adambkaplan) [inline internal/sharedresource/sharedresource.go:34]: Okay for now - may need to change once #43 merges.

File: internal/sharedresource/sharedresource.go
Comment (adambkaplan) [inline internal/sharedresource/sharedresource.go:35]: Along the lines of prior comments - we won't need this if we just add the pull specs of the container images directly, and let Konflux take care of the rest.

File: internal/controller/openshiftbuild_controller.go
Comment (adambkaplan) [inline internal/controller/openshiftbuild_controller.go:52]: Pointer here?

```suggestion
	SharedResource *sharedresource.SharedResource
```

File: internal/controller/openshiftbuild_controller.go
Comment (adambkaplan) [inline internal/controller/openshiftbuild_controller.go:181]: Not blocking:

I think it's totally acceptable to set the value to `Enabled` here. I don't think k8s will allow this due to CRD validations, but it doesn't hurt to ensure a proper value is set here.

**Review by adambkaplan — CHANGES_REQUESTED** [review]
A lot here - the main points boil down to:

1. Use image pull specs in the YAML for shared resources. Ideally use digest shas, but tags are OK.
2. Ensure this PR is rebased on top of #29, then rip out the logic you have to inject images by environment variable/other means.

Everything else can be done in follow up pull requests. Getting code in to unblock others matters more at this point.

File: internal/sharedresource/sharedresource.go
Comment (apoorvajagtap) [reply internal/sharedresource/sharedresource.go:76]: The deleteManifests is invoked if either SharedResource is disabled or the owner is being deleted (in cases when owner is marked for deletion with SharedResoruce still being enabled, we'll need the step to remove the finalizers).
Hence, another check within deleteManifests() for explicit deletion of resources only when disabled.

File: config/sharedresource/node_service.yaml
Comment (apoorvajagtap) [reply config/sharedresource/node_service.yaml:1]: I think it was defined because k8s won't allow applying the yaml with `spec.ports` not set...?

File: internal/common/transformer.go
Comment (apoorvajagtap) [reply internal/common/transformer.go:10]: I couldn't find any transformer in shipwright's code that injectsFinalizer. I doubt if the current shipwright operator injects any finalizers to the resources (couldn't find any finalizers after deploying the current v1.0.1 op as well).

File: internal/sharedresource/sharedresource.go
Comment (adambkaplan) [reply internal/sharedresource/sharedresource.go:74]: I see where you are going. Please add a code comment with this explanation. I have a feeling we are going to revisit this logic in the future, and a comment will help explain to future maintainers why this second check is here.

File: config/sharedresource/node_service.yaml
Comment (adambkaplan) [reply config/sharedresource/node_service.yaml:1]: Okay - then let's keep it for now. Once we have a bundle with all images in it, we can run tests to determine if this service is referenced anywhere.

**Review by adambkaplan — APPROVED** [review]
@apoorvajagtap please squash your commits, then this PR is okay to merge. There's still areas of improvement (as I noted), but we can take these up in separate pull requests.

**Comment by adambkaplan** [comment]
LGTM

**Pull request merged**
